src/hardware/sblaster.cpp

```diff
@@ -294,6 +308,157 @@ static void InitializeSpeakerState()
+			auto it = filter_map.find(filter);
+			if (it != filter_map.end())
+				filter_type = it->second;
+		}
+
+		FilterState filter_state = FilterState::Off;
+		if (filter_type != FilterType::None) {
+			if (filter_prefs.size() > 1 && filter_prefs[1] == "always_on") {
+				filter_state = FilterState::ForcedOn;
+			} else {
+				filter_state = FilterState::On;
+			}
+		}
+		return std::make_tuple(filter_type, filter_state);
+	};
+	using std::tie;
+	tie(sb.sb_filter_type, sb.sb_filter_state) = get_filter_params("sb_filter");
+	tie(sb.opl_filter_type, std::ignore) = get_filter_params("opl_filter");
+}
+
+static void log_filter_config(const char *output_type, FilterType filter)
+{
+	static const std::map<FilterType, std::string> filter_name_map = {
+	        {FilterType::SB1, "Sound Blaster 1.0"},
+	        {FilterType::SB2, "Sound Blaster 2.0"},
+	        {FilterType::SBPro1, "Sound Blaster Pro 1"},
+	        {FilterType::SBPro2, "Sound Blaster Pro 2"},
+	        {FilterType::SB16, "Sound Blaster 16"},
+	};
+
+	if (filter == FilterType::None) {
+		LOG_MSG("%s: %s filter emulation disabled", CardType(), output_type);
+	} else {
+		auto it = filter_name_map.find(filter);
+		if (it != filter_name_map.end()) {
+			auto filter_type = it->second;
+			LOG_MSG("%s: Emulating %s %s output filter",
+			        CardType(),
+			        filter_type.c_str(),
+			        output_type);
+		}
+	}
+}
+
+static void configure_sb_filter()
+{
+	auto set_filter = [](const uint8_t order, const uint16_t cutoff_freq) {
+		sb.chan->ConfigureLowPassFilter(order, cutoff_freq);
+		sb.chan->SetLowPassFilter(sb.sb_filter_state);
+	};
+
+	auto disable_filter = [] { sb.chan->SetLowPassFilter(FilterState::Off); };
+
+	auto enable_zoh_upsampler = [] {
+		constexpr auto dac_rate = 45454;
+		sb.chan->ConfigureZeroOrderHoldUpsampler(dac_rate);
+		sb.chan->EnableZeroOrderHoldUpsampler();
+	};
+
+	switch (sb.sb_filter_type) {
+	case FilterType::None:
+		disable_filter();
+		enable_zoh_upsampler();
+		break;
+
+	case FilterType::SB1:
+		set_filter(2, 3800);
+		enable_zoh_upsampler();
+		break;
+
+	case FilterType::SB2:
+		set_filter(2, 4800);
+		enable_zoh_upsampler();
+		break;
+
+	case FilterType::SBPro1:
+	case FilterType::SBPro2:
+		set_filter(2, 3200);
+		enable_zoh_upsampler();
+		break;
+
+	case FilterType::SB16:
+		// Resampling from the SB channel rate to the mixer rate applies
+		// brickwall filtering at half the SB channel rate, which
+		// perfectly emulates the dynamic brickwall filter of the SB16.
+		disable_filter();
+		sb.chan->EnableZeroOrderHoldUpsampler(false);
+		break;
+	}
+
+	log_filter_config("PCM", sb.sb_filter_type);
+}
+
+static void configure_opl_filter()
+{
+	auto set_filter = [](mixer_channel_t chan,
+	                     const uint8_t order,
+	                     const uint16_t cutoff_freq) {
+		chan->ConfigureLowPassFilter(order, cutoff_freq);
+		chan->SetLowPassFilter(FilterState::On);
+	};
+
+	auto chan = MIXER_FindChannel("FM");
+	if (!chan)
+		return;
```

Comment:
Suggest asserting this; it will catch order-of-operation issues.

Reply:
Is it possible to place this `configure_opl_filter()` function inside the `adlib.cpp` code, where the `"FM"` channel is created?

In `adlib.cpp`, it does have access to the `[sblaster]` ini settings:

``` c++
        mixerChan = MIXER_AddChannel(OPL_CallBack, 0, "FM");

       //  section->Get_string("opl_filter");
       // configure the OPL filter on the channel
```

Adlib is where we want to setup and configure all the OPL related stuff (including its channel).

Reply:
OPL_Init should be the one to configure_opl_filter (from within itself).

Reply:
> Suggest asserting this to catch order of operation bugs in the creation of the Sound Blaster.

Good point, done.

Reply:
> OPL_Init should be the one to configure_opl_filter (from within itself).

Makes sense. I figured you can never use OPL emulation without SB emulation so I thought everything will go into `sblaster.cpp`, but putting it into `adblib.cpp` is more correct, agreed.

Reply:
Actually, no, I think it makes the most sense to leave `configure_opl_filter()` in `sblaster.cpp`. That function uses the `sbtype` to figure out the correct OPL filter to use for each SB model. Because they're so closely tied together, most of the code to determine the PCM and OPL filter types in `configure_filters` is common. To move `configure_opl_filter()` into `adblib.cpp`, we'd need to call `configure_filter` from `adlib.cpp`, make things external, etc. Seems like a lot of complication for not much benefit; after all, you cannot use Adlib/OPL without SB anyway? They're closely related, and like I said, the OPL filter to use is determined by the SB model, not the OPL mode.

Right know `sblaster.cpp` "does things" to the OPL path (e.g. by finding the "FM" channel and calling methods on it). I'd prefer to keep `sblaster.cpp` "know" about OPL, and not make `adlib.cpp` dependent on the SB stuff as well (now it's free from it). That sounds like a circular dependendency, at least in principle. Now `sblaster.cpp` managed both the SB and its embedded OPL implementation, which makes sense, that reflects the hardware being emulated too.

Reply:
> You cannot use Adlib/OPL without SB anyway? 

Hmm interesting.. I was worried about the "Adlib-only" pathway, which is:

``` ini
[sblaster]
sbtype  = none
oplmode = opl2
```

I was thinking this would `OPL_Init()` the OPL code without traversing `SBLASTER_Init()`, and therefore skipping the filter initialization only done in **SBLASTER_Init()**.

But you're absolutely right, even with `sbtype = none`, SBlaster_Init() still stands up the CMS / GameBlaster / OPL routines and their mixer channels before punting out half-way though:

``` c++
		if (sb.type==SBT_NONE || sb.type==SBT_GB) return;
```

OK! 

> Because they're so closely tied together, most of the code to determine the PCM and OPL filter types in configure_filters is common. To move configure_opl_filter() into adblib.cpp, we'd need to call configure_filter from adlib.cpp, make things external,

I was sure how linked this was - and indeed, that would be tedious! (thankfully it's all here in SB anyway, so it makes sense to leverage it all locally) -- thanks (once again) for talking me through :rocket: 

(resolving)

Reply:
> But you're absolutely right, even with `sbtype = none`, SBlaster_Init() still stands up the CMS / GameBlaster / OPL routines and their mixer channels before punting out half-way though

Yeah, they're tightly coupled — and for no good reason, in my opinion. I don't get it why the original authors didn't support emulating the original standalone AdLib card; I don't imagine this would've been super hard to do.

The DOSBox wiki quite accurately describes how it all works:

> DOSBox does not emulate the AdLib Music Synthesizer Card directly. However, it does emulate Sound Blaster cards that feature AdLib support in their synthesizers.

https://www.dosbox.com/wiki/Sound#AdLib